Iter9

repriest opened the pull request:
- extended handler with file storage
- write id-shorturl-longurl as json in file

<!-- GitButler Footer Boundary Top -->
---
This is **part 9 of 14 in a stack** made with GitButler:
- <kbd>&nbsp;14&nbsp;</kbd> #14 
- <kbd>&nbsp;13&nbsp;</kbd> #13 
- <kbd>&nbsp;12&nbsp;</kbd> #12 
- <kbd>&nbsp;11&nbsp;</kbd> #11 
- <kbd>&nbsp;10&nbsp;</kbd> #10 
- <kbd>&nbsp;9&nbsp;</kbd> #9 👈 
- <kbd>&nbsp;8&nbsp;</kbd> #8 
- <kbd>&nbsp;7&nbsp;</kbd> #7 
- <kbd>&nbsp;6&nbsp;</kbd> #6 
- <kbd>&nbsp;5&nbsp;</kbd> #5 
- <kbd>&nbsp;4&nbsp;</kbd> #4 
- <kbd>&nbsp;3&nbsp;</kbd> #3 
- <kbd>&nbsp;2&nbsp;</kbd> #2 
- <kbd>&nbsp;1&nbsp;</kbd> #1 
<!-- GitButler Footer Boundary Bottom -->

Commit:
encodeHandler implemented

Commit:
decodeHandler impemented

+ err shadowing

Commit:
decodeHandler impemented

+ err shadowing

Commit:
redirect header added

Commit:
scheme fixes

Commit:
added location header

Commit:
removed content-type check

Commit:
encode & decode handler tests

Commit:
minor fixes + client

Header().Values[0] instead of Header().Get
Body.Close() in encode handler

Commit:
refactored using chi

changed mux to chi router
added http.error

Commit:
flags introduced

tests fixed for chi router

Commit:
minor fixes

Commit:
added env parsing

Commit:
added middleware logging

- log level passed via args

Commit:
added shortenJson

- /api/shorten handler
- handlers now methods
- restructured project
- removed fmt logging

Commit:
minor fixes

ShortenJsonHandler now ShortenJSONHandler

Commit:
added zipper middleware

- introduced gzip middleware to handlers
- compress/decompress tests

Commit:
append entries to file

- extended handler with file storage
- write id-shorturl-longurl as json in file

Commit:
fixed tests

Perederey commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,97 @@
+	return nil
+}
+
+func (s *Storage) save() error {
```
А вот тут у тебя проблема с производительностью - ты каждый раз перезаписываешь весь файл при добавлении записи. Мы на работе обычно делаем добавление записи в файл, не перезаписывая весь файл.

Perederey commented on cmd/shortener/main.go:
```diff
@@ -1,3 +1,46 @@
+	}
+}
+
+func run() error {
```
Ну тут конечно лучше переделать функцию run() - она делает слишком много всего сразу. Стоит разбить ее на несколько более специализированных функций, чтобы улучшить читаемость и упростить тестирование. Каждая функция должна выполнять одну конкретную задачу, что упростит отладку и поддержку кода в будущем.

Perederey commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,97 @@
+	OriginalURL string `json:"original_url"`
+}
+
+type Storage struct {
```
Storage делает слишком много - и файловые операции, и работу с данными в памяти. Лучше разделить на Repository и FileStorage

Perederey reviewed (commented):
Привет! Хорошо справился с заданием, круто, что разделил код на логические пакеты и покрыл все тестами.

Из того, на что стоит обратить внимание:
- Производительность страдает из-за полной перезаписи файла при каждом добавлении записи
- Много дублированного кода в конфигурации и хэндлерах

А так молодец! Продолжай в том же духе!

Commit:
storage interface

- added storage interface (load, append)
- added repository as in-memory store
- split run() to specific funcs
- added readRequestBody

Commit:
corrected flags
